eip/EIPS/eip-4337.md

```diff
@@ -271,23 +280,27 @@ In order to distinguish between them, there is a call to the NUMBER opcode (`blo
-   A bundler MAY whitelist specific paymaster addresses, and not enforce the above storage limitations
-4. The third (validateUserOpSignature) view-call doesn't access _mutable state_ of any contract except the aggregator AND wallet itself.
-5. The third (validateUserOpSignature) is the first view call after the "NUMBER" marker. If that aggregator is banned, then this staticcall MUST revert immediately, as if it was called with zero gas.
-6. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
-7. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
-8. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` to addresses with `EXTCODESIZE=0`.
-9. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
-10. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
-11. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
-
-Since the wallet is allowed to access its own entry point deposit in order to top it up when needed, the client must know the storage slot in order to whitelist it. The entry point therefore implements the following view function:
-
-```c++
-function getSenderStorage(address sender) external view returns (uint256[] memory senderStorageCells)
-```
+2. The first (validateUserOp) call is allowed to access storage slot of the wallet itself (see Storage access by Slots, below), 
+   but only on contracts that are not part of the current bundle (that is, not in a paymaster or any other wallet)
+   Note that the wallet is allowed (and expected) to deposit to its own balance in the EntryPoint. That is allowed by this rule.
+3. The second (validatePaymasterUserOp), the paymaster may access storage slots of itself AND of the wallet, above.
+   A bundler MAY whitelist specific paymaster addresses, and not enforce this storage access limitation.
+4. The third (validateUserOpSignature) view-call, the aggregator is allowed to access storage slots of itself and of the wallet. 
+5. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
+6. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
+7. No referenced address (`CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL`, `EXTCODESIZE`, `EXTCODEHASH` to addresses with `EXTCODESIZE=0`.
+8. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
+9. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
+10. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
+
+#### Storage access by slots
+
+A contract `A` may reference storage of
```

Comment:
During validation, ...
(yes, it should be clear from the context, but let's make it clear to avoid confusion)